services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionAsyncClient.java

```diff
@@ -15,80 +15,131 @@
+                         return;
+                     }
+                     returnFuture.complete(r);
+                 });
+        return returnFuture;
     }
 
-    //Cannot avoid unchecked cast without upstream changes to supply builder function
-    @SuppressWarnings("unchecked")
-    private <T extends S3Request> T requestWithDecoratedEndpointProvider(T request, String bucket) {
-        return (T) request.toBuilder()
-                          .overrideConfiguration(getOrCreateConfigWithEndpointProvider(request, bucket))
-                          .build();
+    private <T extends S3Request, ReturnT> void requestWithCrossRegion(T request,
+                                                                       Function<T, CompletableFuture<ReturnT>> operation,
+                                                                       String bucketName,
+                                                                       CompletableFuture<ReturnT> returnFuture,
+                                                                       Throwable t) {
+
+        Optional<String> bucketRegionFromException = getBucketRegionFromException((S3Exception) t.getCause());
+        if (bucketRegionFromException.isPresent()) {
+            sendRequestWithRightRegion(request, operation, bucketName, returnFuture,
+                                       bucketRegionFromException);
+        } else {
+            fetchRegionAndSendRequest(request, operation, bucketName, returnFuture);
+        }
     }
 
-    //TODO: optimize shared sync/async code
-    private AwsRequestOverrideConfiguration getOrCreateConfigWithEndpointProvider(S3Request request, String bucket) {
-        AwsRequestOverrideConfiguration requestOverrideConfig =
-            request.overrideConfiguration().orElseGet(() -> AwsRequestOverrideConfiguration.builder().build());
-
-        S3EndpointProvider delegateEndpointProvider = (S3EndpointProvider)
-            requestOverrideConfig.endpointProvider().orElseGet(() -> serviceClientConfiguration().endpointProvider().get());
-
-        return requestOverrideConfig.toBuilder()
-                                    .endpointProvider(BucketEndpointProvider.create(delegateEndpointProvider, bucket))
-                                    .build();
+    private <T extends S3Request, ReturnT> void fetchRegionAndSendRequest(T request,
+                                                                          Function<T, CompletableFuture<ReturnT>> operation,
+                                                                          String bucketName,
+                                                                          CompletableFuture<ReturnT> returnFuture) {
+
+        // // TODO: will fix the casts with separate PR
+        ((S3AsyncClient) delegate()).headBucket(b -> b.bucket(bucketName)).whenComplete((response,
+                                                                                         throwable) -> {
```

Comment:
Can we use the same variables for response/throwable here and further up? I don't mind either r/t or response/throwable, I just think it should be consistent.